internal/cli/cmd/promote/promote.go

```diff
@@ -0,0 +1,129 @@
+package promote
+
+import (
+	"fmt"
+
+	"connectrpc.com/connect"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"k8s.io/utils/ptr"
+
+	typesv1alpha1 "github.com/akuity/kargo/internal/api/types/v1alpha1"
+	"github.com/akuity/kargo/internal/cli/client"
+	"github.com/akuity/kargo/internal/cli/config"
+	"github.com/akuity/kargo/internal/cli/option"
+	v1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+)
+
+func NewCommand(cfg config.CLIConfig, opt *option.Option) *cobra.Command {
+	var freight string
+	var stage string
+	var subscribersOf string
+
+	cmd := &cobra.Command{
+		Use:   "promote [--project=project] --freight=freight-id [--stage=stage] [--subscribers-of=stage]",
+		Short: "Manage the promotion of freight",
+		Args:  option.NoArgs,
+		Example: `
+# Promote a freight to a stage for a specific project
+kargo promote --project=my-project --freight=abc123 --stage=dev
+
+# Promote a freight to subscribers of a stage for a specific project
+kargo promote --project=my-project --freight=abc123 --subscribers-of=dev
+
+# Promote a freight to a stage for the default project
+kargo config set project my-project
+kargo promote --freight=abc123 --stage=dev
+
+# Promote a freight to subscribers of a stage for the default project
+kargo config set project my-project
+kargo promote --freight=abc123 --subscribers-of=dev
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+
+			project := opt.Project
+			if project == "" {
+				return errors.New("project is required")
+			}
+
+			if freight == "" {
+				// TODO: Get latest available freight if empty
+				return errors.New("freight is required")
+			}
+
+			if stage != "" && subscribersOf != "" {
+				return errors.New("stage and subscribers-of can not be supplied simultaneously")
```

Comment:
I wonder a little if the user's mistake is more intuitively understood if we use "--stage and --subscribers-of" in the error message; i.e. with the leading "--".

Note that I don't feel strongly one way or the other. I'm only posing the question. I defer to your judgement. Of course, whichever we select, we should try to honor that convention throughout the CLI. See #1163

Reply:
I felt the same when typing this error message, but left it as an exercise for when flags in the broader sense are refactored. As I feel there is much to gain in how we currently glue them into commands and validate them. But to settle on an approach, I need a broader picture of the other commands.

The primary problem I have at present with this approach as a whole is that they are variables attached to flags, and the error message going into the details of what sets them may make it actually prone to mistakes later on...

Reply:
Ok. So defer to when tackling consistency issues more broadly. sgtm.